### PR TITLE
Issue 9631 - Error message not using fully qualified name (part 2)

### PR DESCRIPTION
**Pull request by ntrel**
@JinShil Following on from #7405.

* Disambiguate incompatible type errors.
* Add "both operands are of type '%s'" error message case.

The latter saves the programmer time reading the error when there are two types with long qualification.

### CODE REVIEW

File: src/dmd/expression.d [hunk 1]
```diff
@@ -4827,11 +4827,19 @@ extern (C++) abstract class BinExp : Expression
         TOK thisOp = (op == TOKquestion) ? TOKcolon : op;
         if (e1.op == TOKtype || e2.op == TOKtype)
         {
-            error("incompatible types for ((%s) %s (%s)): cannot use '%s' with types", e1.toChars(), Token.toChars(thisOp), e2.toChars(), Token.toChars(op));
+            error("incompatible types for ((%s) %s (%s)): cannot use '%s' with types",
+                e1.toChars(), Token.toChars(thisOp), e2.toChars(), Token.toChars(op));
+        }
+        else if (e1.type.equals(e2.type))
+        {
+            error("incompatible types for ((%s) %s (%s)): both operands are of type '%s'",
```
Comment (BBasile) [inline src/dmd/expression.d]: I think it would be better to denote that _the operator %s cant be used for two variable of type %s_.
But maybe in another PR ?

File: src/dmd/expression.d [hunk 2]
```diff
@@ -4827,11 +4827,19 @@ extern (C++) abstract class BinExp : Expression
         TOK thisOp = (op == TOKquestion) ? TOKcolon : op;
         if (e1.op == TOKtype || e2.op == TOKtype)
         {
-            error("incompatible types for ((%s) %s (%s)): cannot use '%s' with types", e1.toChars(), Token.toChars(thisOp), e2.toChars(), Token.toChars(op));
+            error("incompatible types for ((%s) %s (%s)): cannot use '%s' with types",
```
Comment (JinShil) [inline src/dmd/expression.d]: I think we can change `((%s) %s (%s))` to `` `(%s) %s (%s)` ``.  That is, lose the outer parentheses and surround the expression in back ticks.

Comment (JinShil) [reply src/dmd/expression.d]: Also, those single quotes should be backticks.

File: test/fail_compilation/fail3.d [hunk 3]
```diff
@@ -14,21 +14,14 @@ template vector(T)
...
         return r;
     }
-
-    vec2 make2(T x, T y)
-    {
-        vec2 a;
-        a.x = x;
-        a.y = y;
-        return a;
-    }
```
Comment (JinShil) [inline test/fail_compilation/fail3.d]: Why was this removed?

File: test/fail_compilation/bug9631.d [hunk 4]
```diff
@@ -1,21 +1,32 @@
...
 }
 
-template T2()
+void convert()
 {
-    struct F { }
+    F x = tem!().F();
 }
 
-void main()
```
Comment (JinShil) [inline test/fail_compilation/bug9631.d]: Why the refactoring of this test case.  There should only be changes to the error messages.

File: test/fail_compilation/fail3.d (hunk 3 above)
Comment (ntrel) [reply test/fail_compilation/fail3.d]: It doesn't seem relevant to the test - should I restore it?

File: test/fail_compilation/bug9631.d (hunk 4 above)
Comment (ntrel) [reply test/fail_compilation/bug9631.d]: I thought it was a bit simpler whilst still essentially testing the same thing. But I suppose it may be worth having both the two template version and the new changes.

File: src/dmd/expression.d (hunk 2 above)
Comment (ntrel) [reply src/dmd/expression.d]: I agree, but I was just following the existing format for these errors. I'll do as you suggest for the new error added in this PR, but the other ones can be fixed in a separate PR.

File: src/dmd/expression.d (hunk 1 above)
Comment (ntrel) [reply src/dmd/expression.d]: Reworded in f73a72b.

File: test/fail_compilation/bug9631.d (hunk 4 above)
Comment (ntrel) [reply test/fail_compilation/bug9631.d]: Restored the two template version.

File: src/dmd/expression.d [hunk 5]
```diff
@@ -4827,11 +4827,19 @@ extern (C++) abstract class BinExp : Expression
...
+                e1.toChars(), Token.toChars(thisOp), e2.toChars(), Token.toChars(op));
+        }
+        else if (e1.type.equals(e2.type))
+        {
+            error("invalid operation for `(%s) %s (%s)` with both operands of type `%s`",
```
Comment (BBasile) [inline src/dmd/expression.d]: the error message is still bad. My first review comment was more accurate. Maybe:
> "the  `%s` operator is not supported for two operands of type `%s`  (`(%s) %s (%s)`)".

Comment (ntrel) [reply src/dmd/expression.d]: I'm not sure there's much difference, but if necessary I'll update it.

Comment (BBasile) [reply src/dmd/expression.d]: necessary

File: test/fail_compilation/fail3.d (hunk 3 above)
Comment (JinShil) [reply test/fail_compilation/fail3.d]: Yes, it does seem unnecessary, but I don't think it belongs in this PR.

Comment (ntrel) [reply test/fail_compilation/fail3.d]: OK, restored.

**Comment by ntrel** [comment]
I've reverted the rewording of the new error message, so it now just follows the existing format, and does not use backticks. The backtick change should be done in a separate PR as it affects about 8 error messages in dmd, I could do it if this gets merged soon.

File: src/dmd/expression.d (hunk 2 above)
Comment (JinShil) [inline src/dmd/expression.d]: This error message doesn't seem to have a test case.

**Review by JinShil — DISMISSED** [review]

File: src/dmd/expression.d (hunk 2 above)
Comment (ntrel) [reply src/dmd/expression.d]: @JinShil What about `fail_compilation/fail54.d` and `fail_compilation/ice8511.d`?

Comment (JinShil) [reply src/dmd/expression.d]: Sorry, I realize that error message didn't actually change.

**Comment by ntrel** [comment]
Thanks for reviewing.

> The backtick change should be done in a separate PR

See #7464.
